scripts/generate_driver_wrappers.py

```diff
@@ -22,52 +22,117 @@
 
 import sys
 import os
+import json
+from typing import Tuple
 import argparse
+import jsonschema
+from jsonschema import validate
 import jinja2
 from mbedtls_dev import build_tree
 
-def render(template_path: str) -> str:
+def render(template_path: str, driver_jsoncontext: list) -> str:
     """
-    Render template from the input file.
+    Render template from the input file and driver JSON.
     """
     environment = jinja2.Environment(
         loader=jinja2.FileSystemLoader(os.path.dirname(template_path)),
         keep_trailing_newline=True)
     template = environment.get_template(os.path.basename(template_path))
 
-    return template.render()
+    return template.render(drivers=driver_jsoncontext)
 
-def generate_driver_wrapper_file(mbedtls_root: str, output_dir: str) -> None:
+
+def generate_driver_wrapper_file(template_dir: str, \
+                                 output_dir: str, driver_jsoncontext: list) -> None:
     """
     Generate the file psa_crypto_driver_wrapper.c.
     """
     driver_wrapper_template_filename = \
-        os.path.join(mbedtls_root, \
-        "scripts/data_files/driver_templates/psa_crypto_driver_wrappers.c.jinja")
+        os.path.join(template_dir, "psa_crypto_driver_wrappers.c.jinja")
 
-    result = render(driver_wrapper_template_filename)
+    result = render(driver_wrapper_template_filename, driver_jsoncontext)
 
     with open(os.path.join(output_dir, "psa_crypto_driver_wrappers.c"), 'w') as out_file:
         out_file.write(result)
 
+
+def validate_json(driverjson_data: list, driverschema: list) -> bool:
+    """
+    Validate the Driver JSON against schema
+    """
+    try:
+        validate(instance=driverjson_data, schema=driverschema)
+    except jsonschema.exceptions.ValidationError as err:
+        print(err)
+        print("The driver JSON data is InValid")
+        return False
+
+    return True
+
+def merge_driverjsonfiles(mbedtls_root: str, json_directory: str, \
+                          jsondriver_list: str) -> Tuple[bool, list]:
+    """
+    Merge driver JSON files into a single ordered JSON after validation.
+    """
+    result = list()
+    driverlist = list()
+    with open(os.path.join(mbedtls_root, \
+        'scripts/data_files/driver_jsons/driver_transparent_schema.json'), 'r') as file:
+        transparent_driver_schema = json.load(file)
+    with open(os.path.join(mbedtls_root, \
+        'scripts/data_files/driver_jsons/driver_opaque_schema.json'), 'r') as file:
+        opaque_driver_schema = json.load(file)
+
+    with open(os.path.join(json_directory, jsondriver_list), 'r') as driverlistfile:
+        driverlist = json.load(driverlistfile)
+    for file_name in driverlist:
+        with open(os.path.join(json_directory, file_name), 'r') as infile:
+            json_data = json.load(infile)
+            if json_data['type'] == 'transparent':
```

Comment:
If that information is contained already in the `json_data` which is passed in the validate_data it should be proccessed there.

As @gilles-peskine-arm proposed above the list of the schemas should be passed to the function

**reference code**

~~~~~~
def validate_json(driverjson_data: list, driverschema: list) -> bool:
    """
    Validate the Driver JSON against schema according to its 'type' property
    """

   try:
      _schema = driverschema[driverjson_data['type']]
      validate(instance=driverjson_data, schema=_schema)
   except KeyError as e:
      # This could happen if the json_data['type'] does not exist in the passed in schema list
      # schemas = {'transparent': transparent_driver_schema, 'opaque': opaque_driver_schema}
      print(e)
      return False

   except jsonschema.exceptions.ValidationError as err:
   # This exception is triggered when validation fails
      print(err)
      print("The driver JSON data is InValid")
      return False

    return True
~~~~~~

Then as proposed above the code could be simplified to

~~~~~
schemas = {'transparent': transparent_driver_schema, 'opaque': opaque_driver_schema}

# Add error  hanlding for json.load exceptions
_driver_data = [json.load(driver_file_name) for driver_file_name in driver_list]


# Add error handling for validate exceptions
results = [validate(_dt, schemas) for _dt in _driver_data]
~~~~~

Please note that this is reference code, and it may contain syntax errors. Also that the above _var notation is used because it makes it easier to spot newly introduced variables.

Reply:
@minosgalanakis , mypy is not very happy with me indexing the Driver object.
Is there an easy way to make mypy happy about this, I've tried some typing casts that didn't work.

I recollect it seemed to pass when I was doing this after a json load pre review.

scripts/generate_driver_wrappers.py:68: error: Value of type "Driver" is not indexable
scripts/generate_driver_wrappers.py:69: error: Value of type "Driver" is not indexable

Reply:
This has been fixed by creating the Driver as of type dictionary. Considering that there are some gaps WRT MyPy and JSONObjects, this seemed like an okay workaround after trying 'Union' of many types and 'Any'.